tmail-backend/tmail-third-party/open-ai/src/main/java/com/linagora/tmail/mailet/AIBotConfig.java

```diff
@@ -0,0 +1,106 @@
+    public static String MODEL_PARAMETER_NAME = "model";
+    public static String BASE_URL_PARAMETER_NAME = "baseURL";
+
+    public static final LlmModel DEFAULT_LLM_MODEL =
+            new LlmModel("gpt-4o-mini");
+
+    private final String apiKey;
+    private final URL baseURL;
+    private final MailAddress gptAddress;
+    private final LlmModel llmModel;
+
+    public AIBotConfig(String apiKey, MailAddress gptAddress, LlmModel llmModel, @Nullable URL baseURL) {
+        Preconditions.checkNotNull(apiKey);
+        Preconditions.checkNotNull(gptAddress);
+        Preconditions.checkNotNull(llmModel);
+
+        this.apiKey = apiKey;
+        this.baseURL = baseURL;
+        this.gptAddress = gptAddress;
+        this.llmModel = llmModel;
+    }
+
+    public static AIBotConfig fromMailetConfig(MailetConfig mailetConfig) throws MailetException {
+        String apiKeyParam = mailetConfig.getInitParameter(API_KEY_PARAMETER_NAME);
+        String gptAddressParam = mailetConfig.getInitParameter(GPT_ADDRESS_PARAMETER_NAME);
+        String llmModelParam = mailetConfig.getInitParameter(MODEL_PARAMETER_NAME);
+        String baseUrlParam = mailetConfig.getInitParameter(BASE_URL_PARAMETER_NAME);
+
+        if (Strings.isNullOrEmpty(apiKeyParam)) {
+            throw new MailetException("No value for " + API_KEY_PARAMETER_NAME + " parameter was provided.");
+        }
+
+        if (Strings.isNullOrEmpty(gptAddressParam)) {
+            throw new MailetException("No value for " + GPT_ADDRESS_PARAMETER_NAME + " parameter was provided.");
+        }
+
+        Optional<URL> baseURLOpt = Optional.ofNullable(baseUrlParam)
+                .filter(baseUrlString -> !Strings.isNullOrEmpty(baseUrlString))
+                .flatMap(AIBotConfig::baseURLStringToURL);
+
+        LlmModel llmModel = Optional.ofNullable(llmModelParam)
+                .filter(modelString -> !Strings.isNullOrEmpty(modelString))
+                .map(LlmModel::new).orElse(DEFAULT_LLM_MODEL);
+
+        try {
+            return new AIBotConfig(apiKeyParam, new MailAddress(gptAddressParam), llmModel, baseURLOpt.orElse(null));
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    private static Optional<URL> baseURLStringToURL(String baseUrlString) {
+        try {
+            return Optional.of(URI.create(baseUrlString).toURL());
+        } catch (MalformedURLException e) {
+            LOGGER.warn("Invalid LLM API base URL", e);
+        }
+
+        return Optional.empty();
+    }
+
+    public String getApiKey() {
+        return apiKey;
+    }
+
+    public MailAddress getGptAddress() {
+        return gptAddress;
+    }
+
+    public LlmModel getLlmModel() {
+        return llmModel;
+    }
+
+    public Optional<URL> getBaseURL() {
+        if (baseURL == null) {
+            return Optional.empty();
+        } else {
+            return Optional.of(baseURL);
+        }
```

Comment:
Optional.offNullable

BTW why isn't baseUrl an optional itself?

Reply:
It is now.